app/src/main/java/de/westnordost/streetcomplete/screens/main/RecentLocationStore.kt

```diff
@@ -0,0 +1,49 @@
+package de.westnordost.streetcomplete.screens.main
+
+import android.location.Location
+import de.westnordost.streetcomplete.util.math.EARTH_RADIUS
+import kotlin.math.PI
+import kotlin.math.abs
+import kotlin.math.cos
+
+class RecentLocationStore {
+    private val recentLocations: MutableList<Location> = mutableListOf()
+
+    /** returns a sequence of recent locations, with some minimum time and distance from each other */
+    fun get(): Sequence<Location> = synchronized(recentLocations) {
+        var previousLocation: Location? = null
+        recentLocations.asReversed().asSequence().filter {
+            val loc = previousLocation
+            if (loc == null) {
+                previousLocation = it
+                return@filter true
+            }
+            if (abs(it.elapsedRealtimeNanos - loc.elapsedRealtimeNanos) > LOCATION_MIN_TIME_DIFFERENCE_NANOS
+                && !loc.isTooCloseTo(it)
+            ) {
+                previousLocation = it
+                true
+            } else false
+        }
+    }
+
+    fun add(location: Location) = synchronized(recentLocations) {
+        recentLocations.removeAll {
+            it.elapsedRealtimeNanos <= location.elapsedRealtimeNanos - LOCATION_STORE_TIME_NANOS
+        }
+        recentLocations.add(location)
+    }
+}
+
+// ~ 5 times faster than using SphericalEarthMath distanceTo, could be 20 times when using approximate cos
+private fun Location.isTooCloseTo(location: Location): Boolean {
+    // see https://en.wikipedia.org/wiki/Geographical_distance#Spherical_Earth_projected_to_a_plane
+    val dLat = (latitude - location.latitude) * PI / 180.0
+    val dLon = (longitude - location.longitude) * PI / 180.0
+    val cosDLon = cos((latitude + location.latitude) * PI / 180.0 / 2) * dLon // actually cos could be approximated using 1 - x^2 / 2 + x^4 / 24 - x^6 / 720: 4 times faster and good enough
+    val distanceSquared = EARTH_RADIUS * EARTH_RADIUS * (dLat * dLat + cosDLon * cosDLon) // no need for sqrt
+    return distanceSquared < MAX_DISTANCE_TO_ELEMENT_FOR_SURVEY * MAX_DISTANCE_TO_ELEMENT_FOR_SURVEY / 4
+}
```

Comment:
Hmm:

1. Are you sure that this isn't premature optimization, i.e. actually a performance bottleneck? Also, I'd prefer if "FlatEarthMath" (if it is really necessary) wasn't scattered in the code in some private functions but be in a `FlatEarthMath.kt` and subsequently also tested properly (e.g. 180th longitude)

2. (There is actually [`Location.distanceTo`](https://developer.android.com/reference/android/location/Location#distanceTo(android.location.Location)) but it likely does the same as the SphericalEarthMath function.)

Reply:
I only did some tests in Android Studio, here the overall check whether a simple 3 point way is within survey distance is 10% faster than with `LatLon.distanceTo`.
So probably not worth the optimization.

Actually a faster (or approximate) `SphericalEarthMath.angularDistanceToArc` would be much more helpful.

Reply:
I added `FaltEarthMath.kt`, and some tests whether the results are reasonably close to the `SphericalEarthMath` equivalents.
It's tested for 0.03° (up to ~3 km, depending on latitude) and here the results are within 0.5 m of the spherical functions.

Maybe there could be some additional functions that use the approximate distance calculations if the points are close, and the spherical ones otherwise.